lambdas/access_counts/index.py

```diff
@@ -0,0 +1,161 @@
+
+
+athena = boto3.client('athena')
+s3 = boto3.client('s3')
+
+
+def run_query(named_query_id):
+    output = 's3://%s/%s/' % (QUERY_RESULT_BUCKET, QUERY_TEMP_DIR)
+
+    query = athena.get_named_query(NamedQueryId=named_query_id)['NamedQuery']
+    response = athena.start_query_execution(
+        QueryString=query['QueryString'],
+        QueryExecutionContext=dict(Database=query['Database']),
+        ResultConfiguration=dict(OutputLocation=output)
+    )
+    print("Started query:", response)
+
+    execution_id = response['QueryExecutionId']
+
+    return execution_id
+
+
+def wait_for_query(execution_id):
+    while True:
+        response = athena.get_query_execution(QueryExecutionId=execution_id)
+        print("Query status:", response)
+        state = response['QueryExecution']['Status']['State']
+
+        if state == 'RUNNING':
+            pass
+        elif state == 'SUCCEEDED':
+            break
+        elif state == 'FAILED':
+            raise Exception("Query failed! QueryExecutionId=%r" % execution_id)
+        elif state == 'CANCELLED':
+            raise Exception("Query cancelled! QueryExecutionId=%r" % execution_id)
+        else:
+            assert False, "Unexpected state: %s" % state
+
+        time.sleep(5)
+
+
+def get_query_results(execution_id):
+    args = dict(
+        QueryExecutionId=execution_id,
+        MaxResults=MAX_RESULTS
+    )
+
+    header = True
+
+    while True:
+        response = athena.get_query_results(**args)
+        for row in response['ResultSet']['Rows']:
+            if header:
+                # Athena returns the header before the actual results. Sigh.
+                header = False
+            else:
+                yield [value['VarCharValue'] for value in row['Data']]
+
+        next_token = response.get('NextToken')
+        if next_token is None:
+            break
+        else:
+            args.update(
+                NextToken=next_token
+            )
+
+
+def write_counts_csv(fd, data, field_names):
+    writer = csv.writer(fd)
+    writer.writerow(field_names)
+    for key, value in data.items():
+        row = list(key)
+        row.append(json.dumps(value))
+        writer.writerow(row)
+
+
+def handler(event, context):
+    named_packages = defaultdict(list)  # (bucket, hash) -> [(name, version), ...]
+
+    # (eventname, bucket, key) -> {date: count, ...}
+    object_access_counts = defaultdict(lambda: defaultdict(int))
+
+    # (eventname, bucket, package_name) -> {date: count, ...}
+    package_access_counts = defaultdict(lambda: defaultdict(int))
+
+    # (eventname, bucket, package_name, package_version, package_hash) -> {date: count, ...}
+    package_version_access_counts = defaultdict(lambda: defaultdict(int))
+
+    named_packages_execution_id = run_query(NAMED_PACKAGES_QUERY_ID)
+    access_counts_execution_id = run_query(OBJECT_ACCESS_COUNTS_QUERY_ID)
+
+    named_packages_key = '%s/%s.csv' % (QUERY_TEMP_DIR, named_packages_execution_id)
```

Comment:
no need to change this, but i don't understand why we wouldn't use f-strings for `.format`.